File: ios/oplayer/Controller/Gateway/OpenLedger.m
//
//  OpenLedger.m
//  oplayer
//
//  Created by SYALON on 13-11-20.
//
//

#import "OpenLedger.h"
#import "OrgUtils.h"
#import "VCBase.h"

@interface OpenLedger()
{
}

@end

@implementation OpenLedger

- (void)dealloc
{
}

- (WsPromise*)queryCoinList
{
    id api_base = [self.api_config_json objectForKey:@"base"];
    id assets = [self.api_config_json objectForKey:@"assets"];
    assert(assets);
    id exchanges = [self.api_config_json objectForKey:@"exchanges"];
    assert(exchanges);
    id p1 = [OrgUtils asyncFetchUrl:[NSString stringWithFormat:@"%@%@", api_base, assets] args:nil];
    id p2 = [OrgUtils asyncFetchUrl:[NSString stringWithFormat:@"%@%@", api_base, exchanges] args:nil];
    return [WsPromise all:@[p1, p2]];
}

- (NSArray*)processCoinListData:(NSArray*)data_array balanceHash:(NSDictionary*)balanceHash
{
    //  [asset]
    //{
    //    blockchain = bitshares;
    //    code = BTS;
    //    "display_code" = BTS;
    //    "display_name" = Bitshares;
    //    icons =     {
    //        large = "https://s2.coinmarketcap.com/static/img/coins/64x64/463.png";
    //        medium = "https://s2.coinmarketcap.com/static/img/coins/64x64/463.png";
    //        small = "https://s2.coinmarketcap.com/static/img/coins/64x64/463.png";
    //    };
    //    show =     {
    //        "openledger_dex_gateway" = 1;
    //    };
    //},
    
    //  exchange
    //{
    //    amount =     {
    //        destination =         {
    //            max = 50000000;
    //            min = "0.0001";
    //            mul = 0;
    //        };
    //        source =         {
    //            max = 5000000;
    //            min = "1e-08";
    //            mul = 0;
    //        };
    //    };
    //    destination =     {
    //        asset = "OPEN.DOGE";
    //        blockchain = bitshares;
    //    };
    //    fee =     {
    //        source =         {
    //            "fiat_currency" = "<null>";
    //            type = 0;
    //            value = 0;
    //        };
    //    };
    //    icon =     {
    //        destination =         {
    //            large = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //            medium = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //            small = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //        };
    //        source =         {
    //            large = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //            medium = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //            small = "https://s2.coinmarketcap.com/static/img/coins/64x64/74.png";
    //        };
    //    };
    //    id = 2;
    //    limit =     {
    //        source =         {
    //            24h = 49509208;
    //            once = 49509208;
    //        };
    //    };
    //    memo =     {
    //        destination =         {
    //            enabled = 1;
    //            label = Memo;
    //        };
    //        source =         {
    //            enabled = 0;
    //            label = "";
    //        };
    //    };
    //    options =     {
    //        comment = "";
    //        "default_destination_for_source" = 1;
    //        healthy = 1;
    //        "maintenance_reason" = "";
    //        show =         {
    //            "openledger_dex_gateway" = 1;
    //        };
    //        status = 2;
    //    };
    //    rate = 1;
    //    source =     {
    //        asset = DOGE;
    //        blockchain = dogecoin;
    //    };
    //},
    
    id data_assets = [data_array objectAtIndex:0];
    id data_exchanges = [data_array objectAtIndex:1];
    //  任意一个接口不可用都算失败。
    if (!data_assets || [data_assets isKindOfClass:[NSNull class]]){
        return nil;
    }
    if (!data_exchanges || [data_exchanges isKindOfClass:[NSNull class]]){
        return nil;
    }
    
    NSMutableDictionary* deposit_hash = [NSMutableDictionary dictionary];
    NSMutableDictionary* withdraw_hash = [NSMutableDictionary dictionary];
    for (id item in data_exchanges) {
        id src = [item objectForKey:@"source"];
        id dst = [item objectForKey:@"destination"];
        assert(src && dst);
        if (!src || !dst){
            continue;
        }
        if ([[[src objectForKey:@"blockchain"] lowercaseString] isEqualToString:@"bitshares"]){
            //  withdraw: bitshares to others
            assert([src objectForKey:@"asset"]);
            [withdraw_hash setObject:item forKey:[src objectForKey:@"asset"]];
        }else if ([[[dst objectForKey:@"blockchain"] lowercaseString] isEqualToString:@"bitshares"]){
            //  deposit: others to bitshares
            assert([dst objectForKey:@"asset"]);
            [deposit_hash setObject:item forKey:[dst objectForKey:@"asset"]];
        }
    }

    NSMutableArray* result = [NSMutableArray array];
    for (id item in data_assets) {
        if (![[[item objectForKey:@"blockchain"] lowercaseString] isEqualToString:@"bitshares"]){
            continue;
        }
        
        id asset_symbol = [item objectForKey:@"code"];
        assert(asset_symbol);
        if (!asset_symbol){
            continue;
        }
        
        id withdraw_item = [withdraw_hash objectForKey:asset_symbol];
        id deposit_item = [deposit_hash objectForKey:asset_symbol];
        if (!withdraw_item || !deposit_item){
            continue;
        }
        
        //  status: 0 - disabled, 1 - functions in manual mode, 2 - functions in automatic mode
        id deposit_options = [deposit_item objectForKey:@"options"];
        id withdraw_options = [withdraw_item objectForKey:@"options"];
        BOOL enableWithdraw = [[withdraw_options objectForKey:@"healthy"] boolValue] && [[withdraw_options objectForKey:@"status"] integerValue] != 0;
        BOOL enableDeposit = [[deposit_options objectForKey:@"healthy"] boolValue] && [[deposit_options objectForKey:@"status"] integerValue] != 0;
        
        //  细节参考: https://github.com/bitshares/bitshares-ui/pull/2573/commits/8cc40ece6026b24a9becd0bf305b858e6d0d66c5
        id deposit_amount = [[deposit_item objectForKey:@"amount"] objectForKey:@"source"];
        id withdraw_amount = [[withdraw_item objectForKey:@"amount"] objectForKey:@"destination"];
        id deposit_limit = [[deposit_item objectForKey:@"limit"] objectForKey:@"source"];
        id withdraw_limit = [[withdraw_item objectForKey:@"limit"] objectForKey:@"source"];
        
        NSString* symbol = [asset_symbol uppercaseString];
        assert(symbol);
        assert(balanceHash);
        id balance_item = [balanceHash objectForKey:symbol] ?: @{@"iszero":@YES};
        id backingCoin = [[[withdraw_item objectForKey:@"destination"] objectForKey:@"asset"] uppercaseString];

        GatewayAssetItemData* appext = [[GatewayAssetItemData alloc] init];
        appext.enableWithdraw = enableWithdraw;
        appext.enableDeposit = enableDeposit;
        appext.symbol = symbol;
        appext.backSymbol = backingCoin;
        appext.name = item[@"display_name"];
        appext.intermediateAccount = nil;   //  nil
        appext.balance = balance_item;
        appext.depositMinAmount = [self auxValueToNumberString:[deposit_amount objectForKey:@"min"] zero_as_nil:YES];
        appext.withdrawMinAmount = [self auxValueToNumberString:[withdraw_amount objectForKey:@"min"] zero_as_nil:YES];
        appext.withdrawGateFee = [self auxValueToNumberString:[[[withdraw_item objectForKey:@"fee"] objectForKey:@"source"] objectForKey:@"value"]
                                                  zero_as_nil:YES];
        appext.supportMemo = [[[[withdraw_item objectForKey:@"memo"] objectForKey:@"destination"] objectForKey:@"enabled"] boolValue];
        appext.confirm_block_number = nil;
        appext.coinType = symbol;
        appext.backingCoinType = backingCoin;
        appext.withdrawMaxAmountOnce = [self auxMinValue:[withdraw_amount objectForKey:@"max"]
                                                 value02:[withdraw_limit objectForKey:@"once"]
                                             zero_as_nil:YES];
        appext.withdrawMaxAmount24Hours = [self auxValueToNumberString:[withdraw_limit objectForKey:@"24h"] zero_as_nil:YES];
        appext.depositMaxAmountOnce = [self auxMinValue:[deposit_amount objectForKey:@"max"]
                                                value02:[deposit_limit objectForKey:@"once"]
                                            zero_as_nil:YES];
        appext.depositMaxAmount24Hours = [self auxValueToNumberString:[deposit_limit objectForKey:@"once"] zero_as_nil:YES];
        
        appext.open_withdraw_item = withdraw_item;
        appext.open_deposit_item = deposit_item;
        
        id new_item = [item mutableCopy];
        [new_item setObject:appext forKey:@"kAppExt"];
        [result addObject:[new_item copy]];
    }

    return result;
}

- (WsPromise*)requestDepositAddress:(id)item fullAccountData:(id)fullAccountData vc:(VCBase*)vc
{
    assert(item);
    GatewayAssetItemData* appext = [item objectForKey:@"kAppExt"];
    assert(appext);
    id deposit_item = appext.open_deposit_item;
    id exchanges_id = [deposit_item objectForKey:@"id"];
    assert(exchanges_id);
    
    id account_data = [fullAccountData objectForKey:@"account"];
    id outputAddress = [account_data objectForKey:@"name"];
    
    id request_deposit_address_base = [self.api_config_json objectForKey:@"request_deposit_address"];
    id final_url = [NSString stringWithFormat:@"%@%@", self.api_config_json[@"base"], [NSString stringWithFormat:request_deposit_address_base, exchanges_id]];
    
    id args = @{
                @"destination_address":outputAddress,
                @"destination_memo":@""
                };
    
    return [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        [vc showBlockViewWithTitle:NSLocalizedString(@"kTipsBeRequesting", @"请求中...")];
        [[[OrgUtils asyncFetchUrl:final_url args:args] then:(^id(id resp_data) {
            [vc hideBlockView];
            //{
            //    address = xxxaddress;
            //    memo = "<null>";
            //}
            if (resp_data && [resp_data isKindOfClass:[NSDictionary class]]){
                id addr = [resp_data objectForKey:@"address"];
                if (!addr){
                    resolve(NSLocalizedString(@"kVcDWErrTipsRequestDepositAddrFailed", @"获取充币地址异常，请联系网关客服。"));
                    return nil;
                }
                id memo = [resp_data objectForKey:@"memo"];
                if (!memo){
                    memo = [NSNull null];
                }
                id depositItem = @{
                                   @"inputAddress":addr,
                                   @"inputCoinType":[appext.backingCoinType lowercaseString],
                                   @"inputMemo":memo,
                                   @"outputAddress":outputAddress,
                                   @"outputCoinType":[appext.coinType lowercaseString],
                                   };
                resolve(depositItem);
            }else{
                resolve(NSLocalizedString(@"kVcDWErrTipsRequestDepositAddrFailed", @"获取充币地址异常，请联系网关客服。"));
            }
            return nil;
        })] catch:(^id(id error) {
            [vc hideBlockView];
            resolve(NSLocalizedString(@"kVcDWErrTipsRequestDepositAddrFailed", @"获取充币地址异常，请联系网关客服。"));
            return nil;
        })];
    }];
}

/**
 *  验证地址、备注、数量是否有效
 */
- (WsPromise*)checkAddress:(id)item address:(NSString*)address memo:(NSString*)memo amount:(NSString*)amount
{
    //  REMARK：验证地址、备注、数量
    assert(item);
    GatewayAssetItemData* appext = [item objectForKey:@"kAppExt"];
    assert(appext);
    id exchanges_item = appext.open_withdraw_item;
    id exchanges_id = [exchanges_item objectForKey:@"id"];
    assert(exchanges_id);
    
    id validate_method = [self.api_config_json objectForKey:@"validate"];
    id final_url = [NSString stringWithFormat:@"%@%@", self.api_config_json[@"base"], [NSString stringWithFormat:validate_method, exchanges_id]];
    
    id args = @{
                @"amount":amount,
                @"recipient":address,
                @"memo":memo ?: @""
                };
    
    return [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        [[[OrgUtils asyncPostUrl_jsonBody:final_url args:args] then:(^id(id resp_data) {
            if ([[resp_data objectForKey:@"valid_amount"] boolValue] &&
                [[resp_data objectForKey:@"valid_recipient"] boolValue] &&
                [[resp_data objectForKey:@"valid_memo"] boolValue]){
                resolve(@YES);
            }else{
                resolve(@NO);
            }
            return nil;
        })] catch:(^id(id error) {
            resolve(@NO);
            return nil;
        })];
    }];
}

/**
 *  (public) 查询提币网关中间账号以及转账需要备注的memo信息。
 */
- (WsPromise*)queryWithdrawIntermediateAccountAndFinalMemo:(GatewayAssetItemData*)appext
                                                   address:(NSString*)address
                                                      memo:(NSString*)memo
                                   intermediateAccountData:(NSDictionary*)intermediateAccountData
{
    id exchanges_item = appext.open_withdraw_item;
    id exchanges_id = [exchanges_item objectForKey:@"id"];
    assert(exchanges_id);
    
    id request_deposit_address_base = [self.api_config_json objectForKey:@"request_deposit_address"];
    id final_url = [NSString stringWithFormat:@"%@%@", self.api_config_json[@"base"], [NSString stringWithFormat:request_deposit_address_base, exchanges_id]];
    
    id args = @{
                @"destination_address":address,
                @"destination_memo":memo ?: @""
                };
    
    return [WsPromise promise:^(WsResolveHandler resolve, WsRejectHandler reject) {
        [[[OrgUtils asyncFetchUrl:final_url args:args] then:(^id(id resp_data) {
            //{
            //    address = xxxaddress;
            //    memo = "<null>";
            //}
            if (resp_data && [resp_data isKindOfClass:[NSDictionary class]]){
                id addr = [resp_data objectForKey:@"address"];
                id memo = [resp_data objectForKey:@"memo"];
                if (!addr || !memo){
                    resolve(nil);
                    return nil;
                }
                //  继续查询账号信息
                [[[[ChainObjectManager sharedChainObjectManager] queryFullAccountInfo:address] then:(^id(id full_data) {
                    if (!full_data || [full_data isKindOfClass:[NSNull class]])
                    {
                        resolve(nil);
                        return nil;
                    }
                    id depositItem = @{
                                       @"intermediateAccount":addr,
                                       @"finalMemo":memo,
                                       @"intermediateAccountData":full_data,
                                       };
                    resolve(depositItem);
                    return nil;
                })] catch:(^id(id error) {
                    resolve(nil);
                    return nil;
                })];
            }else{
                resolve(nil);
            }
            return nil;
        })] catch:(^id(id error) {
            resolve(nil);
            return nil;
        })];
    }];
}

@end

File: android/app/src/main/java/com/btsplusplus/fowallet/gateway/OpenLedger.kt
package com.btsplusplus.fowallet.gateway

import android.content.Context
import bitshares.*
import com.btsplusplus.fowallet.R
import com.btsplusplus.fowallet.ViewMask
import com.fowallet.walletcore.bts.ChainObjectManager
import org.json.JSONArray
import org.json.JSONObject

class OpenLedger : GatewayBase() {

    override fun queryCoinList(): Promise {
        val api_base = _api_config_json.getString("base")
        val assets = _api_config_json.getString("assets")
        val exchanges = _api_config_json.getString("exchanges")
        val p1 = OrgUtils.asyncJsonGet("$api_base$assets")
        val p2 = OrgUtils.asyncJsonGet("$api_base$exchanges")
        return Promise.all(p1, p2)
    }

    override fun processCoinListData(data_array: JSONArray, balanceHash: JSONObject): JSONArray? {
        val data_assets = data_array.optJSONArray(0)
        val data_exchanges = data_array.optJSONArray(1)
        //  任意一个接口不可用都算失败。
        if (data_assets == null || data_exchanges == null) {
            return null
        }

        val deposit_hash = JSONObject()
        val withdraw_hash = JSONObject()
        data_exchanges.forEach<JSONObject> { it ->
            val item = it!!
            val src = item.optJSONObject("source")
            val dst = item.optJSONObject("destination")
            if (src == null || dst == null) {
                return@forEach
            }
            if (src.getString("blockchain").toLowerCase() == "bitshares") {
                //  withdraw: bitshares to others
                withdraw_hash.put(src.getString("asset"), item)
            } else if (dst.getString("blockchain").toLowerCase() == "bitshares") {
                deposit_hash.put(dst.getString("asset"), item)
            }
        }

        val result = JSONArray()
        data_assets.forEach<JSONObject> { it ->
            val item = it!!

            if (item.getString("blockchain").toLowerCase() != "bitshares") {
                return@forEach
            }

            val asset_symbol = item.optString("code")
            if (asset_symbol == "") {
                return@forEach
            }

            val withdraw_item = withdraw_hash.optJSONObject(asset_symbol)
            val deposit_item = deposit_hash.optJSONObject(asset_symbol)
            if (withdraw_item == null || deposit_item == null) {
                return@forEach
            }

            //  status: 0 - disabled, 1 - functions in manual mode, 2 - functions in automatic mode
            val deposit_options = deposit_item.getJSONObject("options")
            val withdraw_options = withdraw_item.getJSONObject("options")
            val enableWithdraw = withdraw_options.isTrue("healthy") && withdraw_options.getInt("status") != 0
            val enableDeposit = deposit_options.isTrue("healthy") && deposit_options.getInt("status") != 0

            //  细节参考: https://github.com/bitshares/bitshares-ui/pull/2573/commits/8cc40ece6026b24a9becd0bf305b858e6d0d66c5
            val deposit_amount = deposit_item.getJSONObject("amount").getJSONObject("source")
            val withdraw_amount = withdraw_item.getJSONObject("amount").getJSONObject("destination")
            val deposit_limit = deposit_item.getJSONObject("limit").getJSONObject("source")
            val withdraw_limit = withdraw_item.getJSONObject("limit").getJSONObject("source")

            val symbol = asset_symbol.toUpperCase()
            val balance_item = balanceHash.optJSONObject(symbol)
                    ?: jsonObjectfromKVS("iszero", true)

            val backingCoin = withdraw_item.getJSONObject("destination").getString("asset").toUpperCase()

            val appext = GatewayAssetItemData()
            appext.enableWithdraw = enableWithdraw
            appext.enableDeposit = enableDeposit
            appext.symbol = symbol
            appext.backSymbol = backingCoin
            appext.name = item.getString("display_name")
            appext.intermediateAccount = null
            appext.balance = balance_item
            appext.depositMinAmount = auxValueToNumberString(deposit_amount.getString("min"), true)
            appext.withdrawMinAmount = auxValueToNumberString(withdraw_amount.getString("min"), true)
            appext.withdrawGateFee = auxValueToNumberString(withdraw_item.getJSONObject("fee").getJSONObject("source").getString("value"), true)
            appext.supportMemo = withdraw_item.getJSONObject("memo").getJSONObject("destination").isTrue("enabled")
            appext.confirm_block_number = null
            appext.coinType = symbol
            appext.backingCoinType = backingCoin
            appext.withdrawMaxAmountOnce = auxMinValue(withdraw_amount.getString("max"), withdraw_limit.getString("once"), true)
            appext.withdrawMaxAmount24Hours = auxValueToNumberString(withdraw_limit.getString("24h"), true)
            appext.depositMaxAmountOnce = auxMinValue(deposit_amount.getString("max"), deposit_limit.getString("once"), true)
            appext.depositMaxAmount24Hours = auxValueToNumberString(deposit_limit.getString("once"), true)

            appext.open_withdraw_item = withdraw_item
            appext.open_deposit_item = deposit_item

            item.put("kAppExt", appext)
            result.put(item)
        }

        return result
    }

    override fun requestDepositAddress(item: JSONObject, fullAccountData: JSONObject, ctx: Context): Promise {
        val appext = item.get("kAppExt") as GatewayAssetItemData
        val deposit_item = appext.open_deposit_item
        val exchanges_id = deposit_item!!.getString("id")

        val account_data = fullAccountData.getJSONObject("account")
        val outputAddress = account_data.getString("name")

        val request_deposit_address_base = _api_config_json.getString("request_deposit_address")
        val final_url = String.format("%s%s", _api_config_json.getString("base"), String.format(request_deposit_address_base, exchanges_id))

        val args = JSONObject().apply {
            put("destination_address", outputAddress)
            put("destination_memo", "")
        }

        val p = Promise()
        val mask = ViewMask(R.string.kTipsBeRequesting.xmlstring(ctx), ctx)
        mask.show()
        OrgUtils.asyncJsonGet(final_url, args).then {
            mask.dismiss()

            val resp_data = it as? JSONObject
            if (resp_data != null) {
                val addr = resp_data.optString("address")
                if (addr == null) {
                    p.resolve(R.string.kVcDWErrTipsRequestDepositAddrFailed.xmlstring(ctx))
                    return@then null
                }
                val memo = resp_data.opt("memo") as? String
                val depositItem = JSONObject().apply {
                    put("inputAddress", addr)
                    put("inputCoinType", appext.backingCoinType.toLowerCase())
                    if (memo != null) {
                        put("inputMemo", memo)
                    }
                    put("outputAddress", outputAddress)
                    put("outputCoinType", appext.coinType.toLowerCase())
                }
                p.resolve(depositItem)

            } else {
                p.resolve(R.string.kVcDWErrTipsRequestDepositAddrFailed.xmlstring(ctx))
            }
            return@then null
        }.catch {
            mask.dismiss()
            p.resolve(R.string.kVcDWErrTipsRequestDepositAddrFailed.xmlstring(ctx))
        }
        return p
    }

    /**
     *  验证地址、备注、数量是否有效
     */
    override fun checkAddress(item: JSONObject, address: String, memo: String?, amount: String): Promise {
        val appext = item.get("kAppExt") as GatewayAssetItemData
        val exchanges_item = appext.open_withdraw_item
        val exchanges_id = exchanges_item!!.getString("id")

        val validate_method = _api_config_json.getString("validate")
        val final_url = String.format("%s%s", _api_config_json.getString("base"), String.format(validate_method, exchanges_id))

        val args = JSONObject().apply {
            put("amount", amount)
            put("recipient", address)
            put("memo", memo ?: "")
        }

        val p = Promise()

        OrgUtils.asyncPost_jsonBody(final_url, args).then {
            val resp_data = it as JSONObject
            if (resp_data.isTrue("valid_amount") && resp_data.isTrue("valid_recipient") && resp_data.isTrue("valid_memo")) {
                p.resolve(true)
            } else {
                p.resolve(false)
            }
            return@then null
        }.catch {
            p.resolve(false)
        }

        return p
    }

    /**
     *  (public) 查询提币网关中间账号以及转账需要备注的memo信息。
     */
    override fun queryWithdrawIntermediateAccountAndFinalMemo(appext: GatewayAssetItemData, address: String, memo: String?, intermediateAccountData: JSONObject?): Promise {
        val exchanges_item = appext.open_withdraw_item
        val exchanges_id = exchanges_item!!.getString("id")

        val request_deposit_address_base = _api_config_json.getString("request_deposit_address")
        val final_url = String.format("%s%s", _api_config_json.getString("base"), String.format(request_deposit_address_base, exchanges_id))

        val args = JSONObject().apply {
            put("destination_address", address)
            put("destination_memo", memo ?: "")
        }

        val p = Promise()

        OrgUtils.asyncJsonGet(final_url, args).then {
            val resp_data = it as? JSONObject

            if (resp_data != null) {
                val addr = resp_data.optString("address")
                val memo = resp_data.optString("memo")
                if (addr == null || memo == null) {
                    p.resolve(null)
                    return@then null
                }
                //  继续查询账号信息
                ChainObjectManager.sharedChainObjectManager().queryFullAccountInfo(address).then {
                    val full_data = it as? JSONObject
                    if (full_data == null) {
                        p.resolve(null)
                        return@then null
                    }
                    val depositItem = JSONObject().apply {
                        put("intermediateAccount", addr)
                        put("finalMemo", memo)
                        put("intermediateAccountData", full_data)
                    }
                    p.resolve(depositItem)
                    return@then null
                }.catch {
                    p.resolve(null)
                }
            } else {
                p.resolve(null)
            }
            return@then null
        }.catch {
            p.resolve(null)
        }

        return p
    }

}